marconi/test/Spec/Utxo.hs

```diff
@@ -22,27 +30,135 @@ import Cardano.Api qualified as C
+
+genBlockHeader
+  :: Hedgehog.MonadGen m
+  => m C.BlockNo
+  -> m C.SlotNo
+  -> m C.BlockHeader
+genBlockHeader genB genS = do
+  let validByteSizeLength = 32
+  bs <- Gen.bytes(Range.singleton validByteSizeLength)
+  sn <- genS
+  bn <- genB
+  let (hsh :: C.Hash C.BlockHeader) =  fromJust $ C.deserialiseFromRawBytes(C.proxyToAsType Proxy) bs
+  pure (C.BlockHeader sn hsh bn)
+
+genChainPoint
+  :: Hedgehog.MonadGen m
+  => m C.BlockNo
+  -> m C.SlotNo
+  -> m C.ChainPoint
+genChainPoint genB genS = do
+  (C.BlockHeader sn hsh _) <- genBlockHeader genB genS
+  pure $ C.ChainPoint sn hsh
+
+genChainPoint' :: Hedgehog.MonadGen m => m C.ChainPoint
+genChainPoint' = genChainPoint genBlockNo genSlotNo
+
+genEvents :: Gen (Utxo.StorableEvent Utxo.UtxoHandle)
+genEvents = do
+    txs <- Gen.list (Range.linear 2 5)(CGen.genTx C.ShelleyEra)
+    blkNo  <- genBlockNo
+    cp <- genChainPoint'
+    pure . fromJust $ Utxo.getUtxoEvents Nothing txs blkNo cp
+
+
+headOption :: [a] -> Maybe a
+headOption []    = Nothing
+headOption (x:_) = Just x
 
 tests :: TestTree
-tests = testGroup "Marconi.Index.Specs" $
-    [testPropertyNamed
-     "marconi-index utxos event builder" "Spec. Utxo from Cardano.Api.Tx with targetAddresses"
+tests = testGroup "Marconi.Utxo.Indexer.Specs are:" $
+    [
+     testPropertyNamed "marconi-utxo split-by-address property"
+     "Split UtxoEvents by address"
+     eventsAtAddressTest
+    , testPropertyNamed "marconi-utxo event-to-sqlRow property"
+      "Roundtrip UtxoEvent to UtxoRow converion"
+     eventToRowRoundTrip
+    , testPropertyNamed "marconi-utxo event-from-Tx property"
+      "Utxo from Cardano.Api.Tx with targetAddresses"
      txToUtxoTest
-    , testPropertyNamed
-      "marconi-index utxos event at address" "Spec. Filter UtxoEvents for an address"
-      addressFilteredEventsTest
-    , testPropertyNamed "marconi-index utxos in-memory store" "Spec. Save and retreive UtxoEvents"
+    , testPropertyNamed "marconi-utxo storage-roundtrip property"
+      "Roundtrip storage test"
       utxoStorageTest
     ]
 
-addressFilteredEventsTest :: Property
-addressFilteredEventsTest = property $ do
+deriving instance Eq (StorableEvent Utxo.UtxoHandle)
+deriving instance Ord (StorableEvent Utxo.UtxoHandle)
+
+
+equivalentSets :: Eq a => Set a -> Set a -> Bool
+equivalentSets us us' = equivalentLists (Set.toList us) (Set.toList us')
+
+-- | Proves two sets are equivalant, but not identical
+--
+equivalentLists :: Eq a => [a] -> [a] -> Bool
+equivalentLists us us' =
+  length us == length us'
+  &&
+  all (const True) [u `elem` us'| u <- us]
+  &&
+  all (const True) [u `elem` us| u <- us']
+
+-- | UtxoEvents equivalent relationship
+-- Not all utxoEvents have defined `Eq` and `Ord` relationship.
+-- As events are disassembled and reassembled, the Ordering of these sub-parts may change
+-- Therefore we used the Equivalence relationship to show two event are morally the same
```

Comment:
Doesn't this mean we should use `Set` rather than `List` to represent these parts?

Reply:
I don't think we can use List for UTXOs, per  Pluts docs on [EUTXO-model figure 3 and Note 3](chrome-extension://efaidnbmnnnibpcajpcglclefindmkaj/https://cicero.ci.iog.io/api/fact/ffdcd1ee-97e2-4661-936d-4b41086d646a/binary) UTXOs should be List and the TxInput should be `Set`
@raduom, would please take a look and let me know if I understood the Specs correctly?

Reply:
ok, my orig analysis was incorrect. we can use Set. lthought there might be more than one UTXO with the same txId, per specs above, the txIx will be different. So we can continue using Set for the UTXOs